src/core/expr/fexpr_count_countna.cc

```diff
@@ -0,0 +1,171 @@
+//------------------------------------------------------------------------------
+// Copyright 2023 H2O.ai
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+// IN THE SOFTWARE.
+//------------------------------------------------------------------------------
+#include "column/const.h"
+#include "column/latent.h"
+#include "column/countna.h"
+#include "column/count_all_rows.h"
+#include "documentation.h"
+#include "expr/fexpr_func.h"
+#include "expr/eval_context.h"
+#include "expr/workframe.h"
+#include "python/xargs.h"
+#include "stype.h"
+namespace dt {
+namespace expr {
+
+template<bool COUNTNA>
+class FExpr_CountNA : public FExpr_Func {
+  private:
+    ptrExpr arg_;
+
+  public:
+    FExpr_CountNA(ptrExpr &&arg)
+      : arg_(std::move(arg)) {}
+
+    std::string repr() const override {
+      std::string out = COUNTNA? "countna" : "count";
+      out += '(';
+      if (arg_->get_expr_kind() != Kind::None) out += arg_->repr();      
+      out += ')';
+      return out;
+    }
+
+
+    Workframe evaluate_n(EvalContext &ctx) const override {
+      Workframe outputs(ctx);
+      Workframe wf = arg_->evaluate_n(ctx);
+      Groupby gby = ctx.get_groupby();
+      // this covers scenarios where
+      // we dont care about the presence or absence of NAs
+      // we just want the total number of rows
+      bool count_all_rows = arg_->get_expr_kind() == Kind::None;
+
+      if (count_all_rows && !COUNTNA) {
+        Column coli;
+        auto value = static_cast<int64_t>(ctx.nrows());
+        if (gby){
+          coli = Const_ColumnImpl::make_int_column(value, 1, SType::INT64);
+          coli = Column(new Latent_ColumnImpl(new CountAllRows_ColumnImpl(std::move(coli), gby)));
```

Comment:
We need to streamline this and inherit `CountAllRows_ColumnImpl` from
`ReduceNullary_ColumnImpl`. Then you won't need to create a dummy `coli` here. `ReduceNullary_ColumnImpl` could look something like this

```cpp
class ReduceNullary_ColumnImpl : public Virtual_ColumnImpl {
  protected:
    Groupby gby_;

  public:
    ReduceUnary_ColumnImpl(const Groupby& gby, SType stype)
      : Virtual_ColumnImpl(gby.size(), stype),
        gby_(gby)
    {}


    ColumnImpl *clone() const override {
      return new ReduceNullary_ColumnImpl(Groupby(gby_), this->stype());
    }


    size_t n_children() const noexcept override {
      return 0;
    }
};
```

Reply:
is it not simpler to just derive the `CountAllRows_ColumnImpl` from the `Virtual_ColumnImpl` directly? or are there plans to reuse the `ReduceNullary_ColumnImpl` somewhere else?

Reply:
Currently, there is only one function in head_reduce_nullary.cc, that is count0(). So if you want, you can inherit directly from Virtual_ColumnImpl. Inheriting from ReduceNullary_ColumnImpl will just add an indicator that count0() is not going to use any columns as an input.